VM changes for the sensing_of block.  This handles lists properly (by…

picklesrus opened the pull request:
… ignoring them like Scratch2 and makes the attribute menu update based on what was chosen in the target menu.  See issues for context.

### Resolves

Resolves #1762 and resolves LLK/scratch-gui#2389 and resolves LLK/scratch-gui#2390.

towerofnix committed:
Begin work on 'old text' costumes

towerofnix committed:
Merge branch 'develop' of https://github.com/LLK/scratch-vm into old-text-costumes

towerofnix committed:
Remove old dead import file

towerofnix committed:
Merge base/text layers

towerofnix committed:
Fix 'promsie' typo in docs for loadOldTextCostume

towerofnix committed:
Fix eslint errors

towerofnix committed:
Automatically create asset for generated 'flattened' image

towerofnix committed:
Fix eslint whitespace error

fsih committed:
Try to make loadCostume more readable, and revert deserializeCostume to changing the costume object directly

fsih committed:
Get text assets from zip

fsih committed:
Load text layer asset

fsih committed:
Merge branch 'develop' into pull915

fsih committed:
Merge branch 'pull915' into textLayer

fsih committed:
Merge images as part of fetch bitmap

fsih committed:
Why do our lint rules require such weird formatting

fsih committed:
Fix scaling to perform nearest neighbor

fsih committed:
Remove todo

fsih committed:
Rearrange so that if we already have the asset, there's no need to go through promise.all

fsih committed:
Remove mysterious swp file

fsih committed:
Better comment

fsih committed:
Remove swp file

fsih committed:
Comment fixes and instantiate textImageElement only as necessary

fsih committed:
Move reject earlier and add loadedOne comment

kchadha committed:
Add rename and delete functionality for cloud variables. Refactor function name for cloud variable creation API.

picklesrus committed:
VM changes for the sensing_of block.  This handles lists properly (by ignoring them like Scratch2 and makes the attribute menu update based on what was chosen in the target menu.

picklesrus committed:
VM changes for the sensing_of block.  This handles lists properly (by ignoring them like Scratch2 and makes the attribute menu update based on what was chosen in the target menu.

kchadha committed:
Add unit tests for new rename and delete cloud var functionality.

kchadha commented on src/blocks/scratch3_sensing.js:
```diff
@@ -313,12 +313,11 @@ class Scratch3SensingBlocks {
         }
 
-        // Variables
+        // Local variables (not lists).
```
I think the word "Local" here is misleading and should probably just say "Target's variables" or something

picklesrus replied:
Fixed!

kchadha commented on src/engine/blocks.js:
```diff
@@ -563,6 +563,16 @@ class Blocks {
+                    if (block.fields.OBJECT.value === '_stage_') {
+                        this._blocks[block.parent].fields.PROPERTY.value = 'backdrop #';
+                    } else {
+                        this._blocks[block.parent].fields.PROPERTY.value = 'x position';
```
I think the only problem I have with this is that we could decide to change the contents of the menu and maybe add an attribute to the top of the menu or re-arrange the menu and this would still be hardcoded to `x position`. Maybe we can keep track of the sprite menu options and the stage menu options in exported constants in the sensing category file `src/blocks/scratch3_sensing.js` and have the scratch-gui code also use those same constants (we might have to re-export them in the top level virtual machine file as well)? This does seem a bit like over kill though....

picklesrus replied:
I agree that is isn't the best. Trying to prioritize though.. How about I file a tech debt issue for tracking?

kchadha commented on src/engine/runtime.js:
```diff
@@ -597,6 +597,14 @@ class Runtime extends EventEmitter {
     }
 
+    /**
+     * Event name for reporting that block was updated and needs.
```
Incomplete comment?

picklesrus replied:
whoops. Fixed.

kchadha commented on test/unit/blocks_sensing.js:
```diff
@@ -229,6 +229,33 @@ test('loud? boolean', t => {
+    const sensing = new Sensing(rt);
+    const s = new Sprite();
+    const target = new RenderedTarget(s, rt);
+    const thing = {
```
maybe rename this to something more descriptive?

picklesrus replied:
Sorry. Changed to variable.

kchadha commented on test/unit/blocks_sensing.js:
```diff
@@ -229,6 +229,33 @@ test('loud? boolean', t => {
+    target.lookupVariableByNameAndType = () => thing;
+    t.equal(sensing.getAttributeOf({PROPERTY: 'cars'}), 'trucks');
+
+    t.end();
```
I think these tests would be more useful if you mocked out the actual variable and added it to the target (and then called getAttributeOf with both the property and the object set, what do you think?

picklesrus replied:
I'm torn. IIUC, I think doing that means testing the functionality of lookupVariableByNameAndType, which should (at least theoretically) be tested somewhere else.  Maybe it isn't though?  To me, it feels less like a unit test if I do it the way you suggested, but probably would increase our test coverage numbers.

kchadha replied:
Part of what we're testing is whether you're providing things like `lookupVariableByNameAndType` the right arguments inside of `getAttributeOf`. Stubbing out `lookupVariableByNameAndType` doesn't get at that and thus isn't really testing whether the logic inside of `getAttributeOf` is correct. I don't think changing it to remove the `lookupVariableByNameAndType` stub would make this a duplicate test of the `lookupVariableByNameAndType` functionality but rather a test of whether the `getAttributeOf` logic functions correctly as a whole...

IMO, it's still a unit test, just testing `getAttributeOf` as one whole unit. Even if we change the internals of `getAttributeOf` and maybe even stop using `lookupVariableByNameAndType`, I would still expect its unit test to work assuming we're not changing the fundamental functionality of what `getAttributeOf` does.

Thoughts?

kchadha replied:
Now I'm a little torn too haha.

picklesrus replied:
Went back to not mocking lookupVariableByNameAndType. I couldn't add variables with target.createVariable because I couldn't figure out how to have it set a value too so I just smashed it right into the dictionary... But, maybe I'm just missing something.

kchadha replied:
You can set a value after creating the variable by `target.variables[<varId>].value = <thing you want>`

Putting the variable in the dictionary is fine too.

kchadha commented on test/unit/blocks_sensing.js:
```diff
@@ -229,6 +229,33 @@ test('loud? boolean', t => {
+    const stage = new RenderedTarget(s, rt);
+    rt.getTargetForStage = () => stage;
+    stage.lookupVariableByNameAndType = () => null;
+    t.equal(sensing.getAttributeOf({PROPERTY: 'stage'}), 0);
```
The non-existent variable should have a less confusing name because it's not immediately clear what's happening here.

picklesrus replied:
Agreed. Mostly copy/paste error.

kchadha reviewed (changes requested):
I think general approach is reasonable, with a comment about hard coding the menu option names.
However, I'm not sure that's worth changing if it ends up being a lot more work...

Also left some comments about the added unit tests.

picklesrus committed:
fix code review comments

picklesrus committed:
fix code review comments

picklesrus commented:
Did most things, 2 real questions/comments left - one about sharing the menu constants and one about what to mock in the tests.

kchadha committed:
Replace reference to deleted variable with specific variable info needed.

Commit:
Remove catches that do nothing

Commit:
Remove conditional because text layer file name always comes from sb2

kchadha committed:
Merge pull request #1775 from kchadha/cloud-rename-delete

Cloud rename delete

fsih committed:
Merge pull request #1737 from fsih/textLayer

Support text layer in sb2 files

ericrosenbaum committed:
Apply sprite's volume to music extension notes and drums (#1735)

* Use a gain node for volume instead of effects chain

* Use a gain node for volume for drums too

Commit:
Add "project changed" event

And emit it whenever we think the project has changed. Try to not emit it when a change has happened internally that shouldn't affect the serialized project.

This iteration fires the event too frequently, e.g., when switching sprites. This is meant as a simple initial implementation that can be improved.

rschamp committed:
Merge pull request #1779 from rschamp/project-dirty-signal

Add "project changed" event

evhan55 committed:
Adding check for open request before calling reject/resolve.  Resolves #1764.

paulkaplan committed:
Emit a RUNTIME_STARTED event to track if the vm has been started

Commit:
Make setLocale consistently refresh blocks

This change makes the VM's `setLocale` method call `refreshBlocks` even
when not changing the language, though it still skips an unnecessary
call to `formatMessage.setup` in this case.

evhan55 committed:
Merge pull request #1781 from evhan55/extensions/rpc-error

Adding check for open request before calling RPC reject/resolve.

paulkaplan committed:
Force exiting after tests finish

paulkaplan committed:
Merge pull request #1783 from paulkaplan/emit-start-event

Emit a RUNTIME_STARTED event to track if the vm has been started

cwillisf committed:
Merge pull request #1784 from cwillisf/setLocale-always-refreshBlocks

Make setLocale consistently refresh blocks

picklesrus committed:
stop mocking method in the test.

picklesrus committed:
stop mocking method in the test.

paulkaplan committed:
Prevent the runtime from spawning new stepping intervals

paulkaplan committed:
Ensure setCompatibilityMode continues to restart if previously started.

This was previous behavior, but the previous commit which prevents restarts based on _steppingInterval truthiness broke it because clearInterval does not modify the value of _steppingInterval

paulkaplan committed:
Merge pull request #1785 from paulkaplan/no-multi-starts

Prevent the runtime from spawning new stepping intervals

picklesrus committed:
Merge branch 'sensing-of-block-variables' of https://github.com/picklesrus/scratch-vm into sensing-of-block-variables

picklesrus commented:
Closing this in favor of https://github.com/LLK/scratch-vm/pull/1786 because github is being weird about reading the diff :(
